internal/source/containers_image.go

```
@@ -0,0 +1,324 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	catalogdv1alpha1 "github.com/operator-framework/catalogd/api/core/v1alpha1"
+	catalogderrors "github.com/operator-framework/catalogd/internal/errors"
+)
+
+const ConfigDirLabel = "operators.operatorframework.io.index.configs.v1"
+
+type ContainersImageRegistry struct {
+	BaseCachePath string
+	SourceContext *types.SystemContext
+}
+
+func (i *ContainersImageRegistry) Unpack(ctx context.Context, catalog *catalogdv1alpha1.ClusterCatalog) (*Result, error) {
+	l := log.FromContext(ctx)
+
+	if catalog.Spec.Source.Type != catalogdv1alpha1.SourceTypeImage {
+		panic(fmt.Sprintf("programmer error: source type %q is unable to handle specified catalog source type %q", catalogdv1alpha1.SourceTypeImage, catalog.Spec.Source.Type))
+	}
+
+	if catalog.Spec.Source.Image == nil {
+		return nil, catalogderrors.NewUnrecoverable(fmt.Errorf("error parsing catalog, catalog %s has a nil image source", catalog.Name))
+	}
+
+	//////////////////////////////////////////////////////
+	//
+	// Resolve a canonical reference for the image.
+	//
+	//////////////////////////////////////////////////////
+	imgRef, err := reference.ParseNamed(catalog.Spec.Source.Image.Ref)
+	if err != nil {
+		return nil, catalogderrors.NewUnrecoverable(fmt.Errorf("error parsing image reference %q: %w", catalog.Spec.Source.Image.Ref, err))
+	}
+
+	canonicalRef, err := resolveCanonicalRef(ctx, imgRef, i.SourceContext)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving canonical reference: %w", err)
+	}
+
+	//////////////////////////////////////////////////////
+	//
+	// Check if the image is already unpacked. If it is,
+	// return the unpacked directory.
+	//
+	//////////////////////////////////////////////////////
+	unpackPath := i.unpackPath(catalog.Name, canonicalRef.Digest())
+	if unpackStat, err := os.Stat(unpackPath); err == nil {
+		if !unpackStat.IsDir() {
+			return nil, fmt.Errorf("unexpected file at unpack path %q: expected a directory", unpackPath)
```

Should it be unrecoverable?

---

This is a strange one.
1. There's nothing a user can change in their `ClusterCatalog` to get past this.
2. It seems unlikely for this to resolve on its own (how did a file end up here?)

We could panic since we don't expect other writers in this directory. Another option is to delete it and recover that way, but I feel like I'd rather not silently ignore unexpected contents.

So I'm leaning toward panic?

---

Updated to panic.